File: Trabalhos Práticos/tp01/TP01Q12 - RECURSIVO - Palíndromo em C/PalindromoRec.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#define NUMENTRADA 1000
#define TAMLINHA   1000

//metodo para verificar se eh a ultima linha d leitura do arquivo
bool isFim(char *s)
{
    return (strlen(s) >= 3 && s[0] == 'F' && s[1] == 'I' && s[2] == 'M');
}

bool isPalindromo(char* original, int i, int j){
    bool resp = true;

    if(j >= i){
        if(original[i] != original[j]){
            resp = false;
        }
        else{
            resp = isPalindromo(original, i+1, j-1);
        }
    }

    return resp;
}

int main()
{
    // inicio leitura pub.in
    char linha[NUMENTRADA][TAMLINHA];
    int numEntrada = 0;

    do
    {
        fgets(linha[numEntrada], TAMLINHA, stdin);
    } while (isFim(linha[numEntrada++]) == false);
    numEntrada--;
    // fim leitura pub.in
    

    //aqui eu chamo a função isPalindroo para cada uma das entradas do pubin;
    for(int i = 0; i < numEntrada; i++ ){
        if(isPalindromo(linha[i], 0, (strlen(linha[i])-2)) == true){
            printf("SIM\n");
        }else{
            printf("NAO\n");
        }
    }


    return 0;
}

File: Trabalhos Práticos/tp01/TP01Q11 - RECURSIVO - Palíndromo em Java/PalindromoRec.java
public class PalindromoRec {
    

    public static Boolean isPalindromo(String original){
        return isPalindromo(original, 0, (original.length()-1));
    }
    //os parametros i e j servem para eu caminhar pela string pelo inicio e pelo final dela
    public static Boolean isPalindromo(String original, int i, int j){
        
        boolean resp = true;

        if(j >= i){
            if(original.charAt(i) != original.charAt(j)){
                resp = false;
            }
            else{
                resp = isPalindromo(original, i+1, j-1);
            }
        }

        return resp;
    }

    
    
    // Metodo para conferir se a string de entrada eh a ultima
    public static boolean isFim(String s) {
        return (s.length() == 3 && s.charAt(0) == 'F' && s.charAt(1) == 'I' && s.charAt(2) == 'M');
    }

    public static void main(String[] args) {

        // inicio Leitura da entrada padrao
        String[] entrada = new String[1000];
        int numEntrada = 0;

        do {
            entrada[numEntrada] = MyIO.readLine();
        } while (isFim(entrada[numEntrada++]) == false);
        numEntrada--; // Desconsiderar ultima linha contendo a palavra FIM
        // fim Leitura da entrada padrao
        


        for(int i = 0; i < numEntrada; i++){

            if(isPalindromo(entrada[i]) == true){
                MyIO.println("SIM");
            }else{
                MyIO.println("NAO");
            }
            
        }
    }

}
